base/inference.jl

```diff
@@ -3478,6 +3396,17 @@ function reindex_labels!(linfo::LambdaInfo, sv::InferenceState)
     end
 end
 
+@eval function return_type(f::ANY, t::ANY)
+    $(Expr(:meta, :pure))
```

Comment:
Ok, I can live with that. However, I thought the only reason `typeinf` can't be used from `@pure` is that inference might call a `@pure` function, but we already have a special case for that in pure_eval_call, effectively granting an exception for this function.